### PR TITLE
Implemented mechanism to allow themes to disable custom colors

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
When configuring theme support for Gutenberg themes can disable custom colors by setting allow-custom-colors to false.
Custom colors can also be configured in a per-use-case basis by changing allowCustomColors  prop when using ColorPalette component.

Sample theme code for disabling custom colors:
add_theme_support( '`disable-custom-colors`' );

## How Has This Been Tested?
Execute unit tests.
In a theme add the sample code and verify in a paragraph and/or button custom colors get disabled.
Go to one of ColorPalette uses and temporarily set an allowCustomColors prop in a ColorPalette use case, verify that if the prop is explicitly passed the component always respects it.

### CODE REVIEW

**Comment by mtias** [comment]
Maybe `disable-custom-colors` is a better name (default to false).

**Comment by jorgefilipecosta** [comment]
Thank you for the feedback @mtias code was updated.

**Comment by youknowriad** [comment]
@jorgefilipecosta I think we decided to deprecate the name "gutenberg" from theme support, we might want to update this like: `add_theme_support( 'disable-custom-colors')` 

Should we add this to the `themes` doc in the handbook?

**Comment by gziolo** [comment]
> Should we add this to the `themes` doc in the handbook?

Yes, please. This should be a default approach from now on :)

**Comment by aaronjorbin** [comment]
Should this be labeled `Backwards Compatability` since it changes the signature of ColorPalette and anyone using it in a custom block will need to update their block?

**Comment by youknowriad** [comment]
@aaronjorbin As far as I can tell, It's not changing the signature of ColorPalette but I may be wrong?

**Comment by jorgefilipecosta** [comment]
Hi @aaroncampbell thank you for your inputs. I agree with @youknowriad we are not changing the signature of ColorPalette we are adding an optional prop with a default value equal to the old behavior, so every use-case works as before that's why we are not changing any core block in this PR.

**Comment by aaronjorbin** [comment]
@youknowriad it's removing `defaultColors`.  I don't know if anyone is using it right now, but it's possible someone was using that instead of `colors` in their block.

**Comment by youknowriad** [comment]
@aaroncampbell If you take a look at https://github.com/WordPress/gutenberg/pull/4991/files#diff-aae07a66df1640e718b1d0b1cea954c4R82 defaultColors were coming for the settings. The check has just been moved to the wrapper function instead of doing in the inner component.

**Comment by jorgefilipecosta** [comment]
Yes @aaronjorbin, you are right I tottaly forgot that refactor, we can revert that change. But I think that this change improves the component. I will add a warning message when the prop is used.

**Comment by aaronjorbin** [comment]
`anyone using it in a custom block will need to update their block`  would have been better worded as `someone using it in a custom block may need to update their block`. The removal of the `should render a dynamic toolbar with default colors, when colors are not present` test is what makes me think this is a breaking change. I could be wrong. 

I'm not saying this shouldn't happen.  Until this has been merged into core, the API shouldn't be considered locked by any means.  I just want to make sure that any changes are called out.

Also, @aaroncampbell and I are not the same person :) Though I'm fully ok with trolling his mentions until he changes his name. :trollface:

**Comment by jorgefilipecosta** [comment]
> anyone using it in a custom block will need to update their block would have been better worded as someone using it in a custom block may need to update their block. The removal of the should render a dynamic toolbar with default colors, when colors are not present test is what makes me think this is a breaking change. I could be wrong.

The test tested a simple behavior that was happening inside the component, the logic to switch between context colors and colors passed is now done in the withContext, so this logic is now not part of the innerComponent and that's why the test was removed.

The only case where defaultColors could be used is if someone imported the inner non-default component that was just exported for testing. At the first, I thought you referring that component but I was probably wrong in my interpretation.

So in summary, as @youknowriad pointed out there is no behavior change when using the default export because even if defaultColors was passed it would be ignored and replaced with the context.

**Comment by aaronjorbin** [comment]
> The only case where defaultColors could be used is if someone imported the inner non-default component that was just exported for testing. At the first, I thought you referring that component but I was probably wrong in my interpretation.

That is in fact what I was refering to. If it's available, someone will find a way to use it, even if they shouldn't.

My concern is seeming more and more minor as we discuss it though it does make me wonder if we should do something so that things which are only exported for testing purposes aren't used for non testing purposes to make it easier to make changes such as these. I'll give that some thought and potentially open a new ticket.

Thanks for working on this.

**Comment by jorgefilipecosta** [comment]
> I think it's not exported outside of the module itself which means no third-party code can access this variable. (same for all components exported for test purpose), so this is not an issue here.

That's awesome news @youknowriad, thank you a lot for the clarification, and sorry the misunderstanding caused because I was not aware of that.

File: docs/themes.md [hunk 1]
```diff
@@ -43,3 +43,12 @@ add_theme_support( 'editor-color-palette',
 ```
 
 The colors will be shown in order on the palette, and there's no limit to how many can be specified.
+
+### Disabling custom colors in block Color Palettes
+
+By default, the color pallet offered to blocks, allows the user to select a custom color different from the editor or theme default colors.
```
Comment (gziolo) [inline docs/themes.md]: Should be:

color pa*l*et**te**

Assuming that the component is properly named :)

File: docs/themes.md [hunk 2]
```diff
@@ -43,3 +43,12 @@ add_theme_support( 'editor-color-palette',
...
+Themes can disable this feature using:
+```php
+add_theme_support( 'disable-custom-colors', true );
+```
+This flag will make sure users are only able to choose colors from the editor-color-pallet the theme provided or from the editor default colors if the theme did not provide one.
```
Comment (gziolo) [inline docs/themes.md]: Can we wrap editor-color-pallet with '`' to better stand out?

File: lib/client-assets.php [hunk 3]
```diff
@@ -880,6 +881,10 @@ function gutenberg_editor_scripts_and_styles( $hook ) {
 		'titlePlaceholder'   => apply_filters( 'enter_title_here', __( 'Add title', 'gutenberg' ), $post ),
 	);
 
+	if ( ! empty( $disable_custom_colors ) && is_bool( $disable_custom_colors[0] ) ) {
```
Comment (gziolo) [inline lib/client-assets.php]: `if ( ! empty( $disable_custom_colors[0] )` - is enough, `empty` operator performs all required checks.
By the way, why this is an array?

File: docs/themes.md (hunk 2 above)
Comment (jorgefilipecosta) [reply docs/themes.md]: In order to create this stand out effect I had to remove ```php wrapping but as it is just for a flag I think it worths it.

File: lib/client-assets.php (hunk 3 above)
Comment (jorgefilipecosta) [reply lib/client-assets.php]: This code was simplified we now use the format add_theme_support( 'disable-custom-colors' ); instead of add_theme_support( 'disable-custom-colors', true ); so no arrays are involved and we are certain we always have a true or false value.

File: docs/themes.md (hunk 1 above)
Comment (jorgefilipecosta) [reply docs/themes.md]: Nice catch, the description was updated.

File: docs/themes.md (hunk 2 above)
Comment (gziolo) [reply docs/themes.md]: You didn't understand me correctly :)

I wanted only this:

> ```php
> add_theme_support( 'disable-custom-colors', true );
> ```
> This flag will make sure users are only able to choose colors from the `editor-color-palette` the theme provided or from the editor default colors if the theme did not provide one.

**Review by gziolo — APPROVED** [review]
I haven't tested this one, but code wise it looks good. There is still one small correction possible for docs.

File: docs/themes.md (hunk 2 above)
Comment (jorgefilipecosta) [reply docs/themes.md]: Ah, sorry the misunderstanding, it was updated.

**Comment by jorgefilipecosta** [comment]
Hi @karmatosed, from the user experience perspective do you think this PR is ready?

**Comment by gziolo** [comment]
@jorgefilipecosta I double check with @karmatosed, it is good to go 👍
